File: Data Structures & Algorithms/target-sum/submission-11.cs
public class Solution {
    public int FindTargetSumWays(int[] nums, int target) {
        Dictionary<int, int> dp = new Dictionary<int, int>();
        dp[0] = 1;

        foreach (int num in nums) {
            Dictionary<int, int> nextDp = new Dictionary<int, int>();
            foreach (var entry in dp) {
                int total = entry.Key;
                int count = entry.Value;

                if (!nextDp.ContainsKey(total + num)) {
                    nextDp[total + num] = 0;
                }
                nextDp[total + num] += count;

                if (!nextDp.ContainsKey(total - num)) {
                    nextDp[total - num] = 0;
                }
                nextDp[total - num] += count;
            }
            dp = nextDp;
        }

        return dp.ContainsKey(target) ? dp[target] : 0;
    }
}

File: Data Structures & Algorithms/two-integer-sum-ii/submission-11.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        l, r = 0, len(numbers) - 1

        while l < r:
            curr = numbers[r] + numbers[l]
            if curr < target:
                l += 1
            elif curr > target:
                r -= 1
            else:
                return [l + 1, r + 1]
